architecture/2024-08-bicepdeployment-controller.md

```diff
@@ -0,0 +1,446 @@
+# BicepDeployment Controller
+
+* **Author**: Will Smith (@willdavsmith)
+
+## Overview
+
+<!--
+Provide a succinct high-level description of the component or feature and 
+where/how it fits in the big picture. The overview should be one to three 
+paragraphs long and should be understandable by someone outside the Radius
+team. Do not provide the design details in this, section - there is a
+dedicated section for that later in the document.
+-->
+
+Today, users of Radius (and importantly, future users of Radius) use many Kubernetes-specific tools in their production workflows, such as Flux and ArgoCD. The usage of these tools presents an issue when trying to integrate Radius - they can't deploy Radius resources directly since Radius resources are not Kubernetes objects. This design document proposes updates to Radius will allow users to deploy Radius resources by authoring Kubernetes CRDs, which will enable usage of Radius with GitOps tools as well as other Kubernetes-specific tools.
+
+## Terms and definitions
+
+<!--
+Include any terms, definitions, or acronyms that are used in
+this design document to assist the reader. They may or may not
+be part of the user-facing experience once implemented, and can
+be specific to this design context.
+-->
+
+**CRD (Custom Resource Definition)**: A Kubernetes resource that allows users to define their own resource types.
+
+**GitOps** A way to do Kubernetes cluster management and application delivery. It works by using Git as a single source of truth for declarative infrastructure and applications.
+
+**Flux**: A GitOps tool that automates the deployment of applications to Kubernetes clusters.
+
+**ArgoCD**: Another GitOps tool that automates the deployment of applications to Kubernetes clusters.
+
+## Objectives
+
+<!--
+Describe goals/non-goals and user-scenario of this feature to understand
+the end-user goals.
+* If the feature shares the same objectives of the existing design, link
+  to the existing doc and section rather than repeat the same context.
+* If the feature has a scenario, UX, or other product feature design doc,
+  link it here and summarize the important parts.
+-->
+
+### Goals
+
+<!--
+Describe goals to define why we are doing this work, how we will make
+priority decisions, and how we will determine success.
+-->
+
+**Goal: Users can use Kubernetes GitOps tools (Flux, ArgoCD) to deploy and manage resources defined in Bicep manifests**
```

Comment:
do we feel like this is feasible? are there any edge cases with Bicep-defined resources where we feel like we would not be able to manage them all with one CRD?

Reply:
For the Bicep controller, users can use **any** Kubernetes tooling. GitOps is our motivation for the project overall, but for this component we don't have a bias towards any specific tool. eg: Helm, Kubectl, Kustomize, KeptN, etc. 

Kubernetes is an API, and adding a controller for Bicep allows the API to manage anything Bicep can manage.

Reply:
> are there any edge cases with Bicep-defined resources where we feel like we would not be able to manage them all with one CRD?

I'm not worried about this. Goal goal.

Useful thought experiment... if there were Bicep things that can't work - would be a design issue or "just" a bug?